src/rdrand.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation for SGX using RDRAND instruction
+use crate::Error;
+use core::arch::x86_64::_rdrand64_step;
+use core::mem;
+use core::num::NonZeroU32;
+
+#[cfg(not(target_feature = "rdrand"))]
+compile_error!("enable rdrand target feature!");
+
+// Recomendation from "Intel® Digital Random Number Generator (DRNG) Software
+// Implementation Guide" - Section 5.2.1 and "Intel® 64 and IA-32 Architectures
+// Software Developer’s Manual" - Volume 1 - Section 7.3.17.1.
+const RETRY_LIMIT: usize = 10;
+
+fn get_rand_u64(data: &mut u64) -> Result<(), Error> {
+    for _ in 0..RETRY_LIMIT {
+        unsafe {
+            if _rdrand64_step(data) == 1 {
+                return Ok(());
+            }
+        };
+    }
+    Err(Error::UNKNOWN)
+}
+
+fn fill_small_slice(slice: &mut [u8]) -> Result<(), Error> {
+    if slice.len() > 0 {
+        let mut data = unsafe { mem::uninitialized() };
+        get_rand_u64(&mut data)?;
+        slice.copy_from_slice(&data.to_ne_bytes()[..slice.len()])
```

Comment:
`copy_from_slice` results in `memcpy` call instead of a simple move. Maybe keep the old `ptr::write_unaligned`?

Reply:
Never mind, we can't use `write_unaligned` for the second `fill_small_slice`, as we will overwrite data outside of the provided slice.

Reply:
BTW this line results in "cannot borrow `*slice` as immutable because it is also borrowed as mutable" compilation error on Rust 1.34.

Reply:
Ya it looks like I was implicitly relying on NLL, should be fixed now. There's no way to implement this without a call to `memcpy`, as the provided slice might be smaller than 8 bytes.